### PR TITLE
feat: implement auto-instrumentation for `fs`

### PR DESCRIPTION
**Pull request by Rauno56**
## Which problem is this PR solving?

This implements partial auto-instrumentation for the native `fs` module.

## Short description of the changes

This instruments most of the `fs` functions. Full list in `constants.ts` file.
Provides two hooks:

- `createHook: (functionName: string, info: { args: ArrayLike<unknown> }) => boolean;`
- `endHook: (functionName: string, info: { args: ArrayLike<unknown>; span: api.Span }) => void;`

Returning `false` from `createHook` skips creating the trace.

## Checklist

- [x] ~Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable.~

### CODE REVIEW

File: plugins/node/instrumentation-fs/src/types.ts
Comment (vmarchaud) [inline plugins/node/instrumentation-fs/src/types.ts:31]: I believe you could use `FMember | FPMember` to have better typechecking ?

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (vmarchaud) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:140]: not sure which nested calls you are refering to ?

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:38]: It's preferable to not use use default value.
[See here](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/892#discussion_r805390997)

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:279]: remove comment?

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:275]: try / catch to protect against exception?

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:291]: I think we should. We should at least be consistent with the case above when `createHook` return false

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:314]: WDYT about also adding the err here to the info? I imagine it can be useful sometimes

Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:314]: Should also be wrapped in try / catch. Otherwise the hook error will be the one returned to the user in case it throws

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:305]: a bit safer alternative would be to place it in a finally block IMO

File: plugins/node/instrumentation-fs/src/types.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/types.ts:34]: I think in other places we usually set the first param to `span` and place function data in the second object

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:163]: try / catch this hook

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:154]: suggestion: move to finally block

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:37]: should we also add it to the `auto-instrument` package?

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:171]: I would suggest naming it `_patchCallbackApiFunction`. When I read `Async` I think promises. This is also the name used in nodejs docs

File: plugins/node/instrumentation-fs/src/constants.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/src/constants.ts:87]: is there a reason not to implement it nevertheless?

File: plugins/node/instrumentation-fs/test/index.test.ts
Comment (blumamir) [inline plugins/node/instrumentation-fs/test/index.test.ts:34]: in `instrumentation.ts` the check for > 8.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:140]: `fs.writeFile` for example calls `open` and `close` internally. Latter are the nested calls I'm referring to.

File: plugins/node/instrumentation-fs/src/types.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/types.ts:31]: :+1: Removed all references to strings from the implmeentation.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:35]: I would use `semver` for anything more complex than that but consciously did not here. Do you think we should?

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:38]: :+1:

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:279]: Done.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:35]: I'm good with both options. Just think that using `semver` is more robust and readable. But I agree it's a bummer to import another package for such a simple task.
Feel free to resolve or change :)

File: plugins/node/instrumentation-fs/src/index.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/index.ts:20]: Absolutely, added exports for member, hook and config types.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:275]: It's a bug here, yes. I refactored running hooks altogether and brought running those into one place and making them safe to run in one go.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:291]: Hmm... interesting. I don't see a strong reason to seek consistency with the `createHook` behavior since the scenarios are different.

1. The question in the comment concerns whether an user can expect to see only top-level calls which were initiated in the top level.
2. `createHook` behavior concerns the use-case when user has disabled the main call thus already signalling disinterest to the calls nested IMHO.

But you are right pointing out that `createHook` doesn't cover all the usecases: for example when I want to catch ALL(even nested) the "open" calls but nothing else.

I'd vote for
1. 70% removing suppressing nested calls and giving that control(and responsibility) to the user OR
2. 30% leaving it as is making it convenient to ignore noise.

I would want to avoid, however making the solution more complex than it absolutely has to be now and in the future. `fs` is very frequently called API and I would want to stay as slim as possible in such a hot path.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:314]: Yes. Addressed that in the hook refactor.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:163]: Addressed in hook refactor.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:314]: Added `error`.

File: plugins/node/instrumentation-fs/src/types.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/types.ts:34]: I've prioritized consistency between the hooks over consistency between the other instrumentations here. But it does make sense to have Span up front for `endHook` - We have it the other way around as well(`responseHook` in aws sdk). Span is not available for `createHook`.

This would be an option:

```typescript
export type CreateHook = (
  info: { functionName: FMember | FPMember; args: ArrayLike<unknown> }
) => boolean;
export type EndHook = (
  span: api.Span,
  info: { functionName: FMember | FPMember; args: ArrayLike<unknown>; error?: Error }
) => void;
```

OR split the info object up to separate arguments altogether. I'd support the latter I think if we're going to move span to be the first argument.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:171]: Good idea.. Changed that everywhere.

File: plugins/node/instrumentation-fs/src/constants.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/constants.ts:87]: No good one.
Just wanted to get the core of it done and included everything that came "free" for the first version of it.

File: plugins/node/instrumentation-fs/test/index.test.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/test/index.test.ts:34]: Unified those.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:305]: Changed to use `finally` for ending spans now. Running `endHook` cannot be moved to `finally` because we're now passing error to it.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:37]: I'd do it separately since packages take some time to get published.

**Comment by Rauno56** [comment]
Thanks **a lot** @vmarchaud and @blumamir. Really valuable feedback! I addressed everything in one way or another and took the liberty to resolve all the threads I found trivial and resolved.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (vmarchaud) [inline plugins/node/instrumentation-fs/src/instrumentation.ts:76]: since some might be not available i think you should check that they are defined before wrapping them, it might be possible that user rely on the function existing or not for their apps

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (dyladan) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:143]: Personally I would. Maybe we can introduce a verbose mode later if users ask for it

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (Rauno56) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:76]: `shimmer` does that for us already and exits early. Do you think I should double-check it here before calling wrap?

> it might be possible that user rely on the function existing or not for their apps

The instrumentation doesn't change the API. I don't think I understood you correctly here, but if the user relies on `rm` function existing(added in v14) for example, but runs v12 we can't and should not do anything about it.

**Review by vmarchaud — APPROVED** [review]

File: plugins/node/instrumentation-fs/package.json
Comment (dyladan) [inline plugins/node/instrumentation-fs/package.json:3]: Why starting at version 27?

File: plugins/node/instrumentation-fs/package.json
Comment (dyladan) [inline plugins/node/instrumentation-fs/package.json:45]: may want to update to latest deps

File: plugins/node/instrumentation-fs/src/constants.ts
Comment (dyladan) [reply plugins/node/instrumentation-fs/src/constants.ts:87]: Thinks its fine to get something out and iterate

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (dyladan) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:143]: I'm going to let you make a decision here as i can see it both ways. As I said I would lean towards suppressing and we can always unsuppress or add a verbose mode later. Hard to go back the other way without breaking alerts and stuff.

**Review by dyladan — APPROVED** [review]
I'll let you decide on the suppressing. I'm going to approve, but not merge until you've made that decision.

File: plugins/node/instrumentation-fs/src/instrumentation.ts
Comment (blumamir) [reply plugins/node/instrumentation-fs/src/instrumentation.ts:43]: 👍

File: plugins/node/instrumentation-fs/src/constants.ts
Comment (blumamir) [reply plugins/node/instrumentation-fs/src/constants.ts:87]: 👍

File: plugins/node/instrumentation-fs/package.json
Comment (blumamir) [reply plugins/node/instrumentation-fs/package.json:45]: If the user uses an old 1.x.x version, this will force him to upgrade to a higher api version in order to use the instrumentation, when in practice older versions will work fine as well, right?
I find myself sometimes wondering if the version should be the minimum supported one (1.0.0), or the latest 🤔

**Review by blumamir — APPROVED** [review]
LGTM, great work. Was waiting for instrumentation for this package, thanks for taking it 🤩

File: release-please-config.json
Comment (blumamir) [inline release-please-config.json:19]: based on #921, should it also be added to `.release-please-manifest.json`?

**Comment by dyladan** [comment]
@Rauno56 you can merge this if you're happy with it. I think with release-please updated we should no longer need to manually add it to the manifest.

File: release-please-config.json
Comment (dyladan) [reply release-please-config.json:19]: I think we don't need that now that release-please is updated to v3

File: plugins/node/instrumentation-fs/package.json
Comment (Rauno56) [reply plugins/node/instrumentation-fs/package.json:3]: Out of a habit. We used to version-match packages. I guess it can now be `0.1.0`, which makes a lot more sense.

File: plugins/node/instrumentation-fs/package.json
Comment (Rauno56) [reply plugins/node/instrumentation-fs/package.json:45]: It makes sense for the implementors/libs to specify the smallest version supported to maximize compatibility across libs.

In that sense, having a core package expose interfaces to implement messes with the semver expectations.

But the I don't see a reason to hold back updating the dev deps.

File: release-please-config.json
Comment (Rauno56) [reply release-please-config.json:19]: Removed that change.

**Comment by Rauno56** [comment]
> Hard to go back the other way without breaking alerts and stuff.

I think that's a good point. I'll change that to be strict for now.

Note to myself: test to make sure user calls in callbacks aren't suppressed.

File: plugins/node/instrumentation-fs/package.json
Comment (Flarna) [reply plugins/node/instrumentation-fs/package.json]: As far as I know renovate bot will create a PR to pin it in case you would use e.g. `^1.0.2`. But on the other hand it seems renovate PRs are mostly ignored anyway meanwhile otherwise we would have 1.0.4 at other places.

Comment (blumamir) [reply plugins/node/instrumentation-fs/package.json]: > But the I don't see a reason to hold back updating the dev deps.

I think CI enforces that the version of `@opentelemetry/api` in peerDependencies matches the version in devDependencies

Comment (Rauno56) [reply plugins/node/instrumentation-fs/package.json]: That's true. I'll address that separately in #936 and update all the packages after that's merged. For now. I'll comply with the restrictions peer-check currently enforces.

> As far as I know renovate bot will create a PR to pin it in case you would use e.g. ^1.0.2. But on the other hand it seems renovate PRs are mostly ignored anyway meanwhile otherwise we would have 1.0.4 at other places.

I agree we could be better with updating the minor versions of packages. It's currently problematic because renovate bot bundles them all together and some of them fail tests - #618

Probably not a huge deal, but requires a significant contribution to resolve.

**Pull request merged**
